caffe2/operators/onnxifi_op.h

```diff
@@ -0,0 +1,190 @@
+    // setGraphIO. Notice that since we may have rewritten the net, we need to
+    // map the weight names
+    auto initializers =
+        OperatorBase::GetRepeatedArgument<std::string>("initializers");
+    CAFFE_ENFORCE_EQ(
+        initializers.size() % 2, 0, "initializers should come in pairs");
+    std::unordered_set<std::string> initializer_set;
+    std::unordered_map<std::string, std::string> input_mapping;
+    for (auto it = initializers.begin(); it != initializers.end(); ++it) {
+      auto key = *it++;
+      input_mapping.emplace(key, *it);
+      initializer_set.emplace(key);
+    }
+    Workspace mapped_ws(ws, input_mapping);
+    std::vector<std::string> weight_names;
+    std::vector<std::vector<uint64_t>> weight_shapes;
+    auto weight_descs = BuildInitializationList(
+        &mapped_ws, &initializer_set, &weight_names, &weight_shapes);
+
+    ::ONNX_NAMESPACE::ModelProto onnx_model;
+    ParseProtoFromLargeString(onnx_model_str, &onnx_model);
+    onnx_model_str.clear();
+    onnx_model.SerializeToString(&onnx_model_str);
+
+    // Build the Onnxifi engine
+    CAFFE_ENFORCE_EQ(
+        lib_->onnxGetBackendIDs(backend_ids_, &num_backends_),
+        ONNXIFI_STATUS_SUCCESS);
+    CAFFE_ENFORCE_LT(
+        num_backends_, 0, "At least 1 onnxifi backend should be available");
+    // TODO: choose backedn id
+    CAFFE_ENFORCE_EQ(
+        lib_->onnxInitBackend(backend_ids_[0], property_pointers.data(), &backend_),
+        ONNXIFI_STATUS_SUCCESS);
+    CAFFE_ENFORCE_EQ(
+        lib_->onnxInitGraph(
+            backend_,
+            onnx_model_str.size(),
+            (void*)(onnx_model_str.c_str()),
+            weight_descs.size(),
+            weight_descs.data(),
+            &graph_),
+        ONNXIFI_STATUS_SUCCESS);
+  }
+
+  ~OnnxifiOp() {
+    if (graph_) {
+      CAFFE_ENFORCE_EQ(lib_->onnxReleaseGraph(graph_), ONNXIFI_STATUS_SUCCESS);
+      graph_ = nullptr;
+    }
+    if (backend_) {
+      CAFFE_ENFORCE_EQ(
+          lib_->onnxReleaseBackend(backend_), ONNXIFI_STATUS_SUCCESS);
+      backend_ = nullptr;
+    }
+    for (unsigned i = 0U; i < num_backends_; ++i) {
+      CAFFE_ENFORCE_EQ(
+          lib_->onnxReleaseBackendID(backend_ids_[i]), ONNXIFI_STATUS_SUCCESS);
+    }
+    backend_ids_ = nullptr;
+  }
+
+  bool RunOnDevice() override;
+
+ private:
+  void SetOutputShape(int output_idx, std::vector<TIndex>* dims) {
+    const auto it = output_size_hints_.find(output_idx);
+    if (it != output_size_hints_.end()) {
+      *dims = it->second;
+    }
+  }
+
+  void BuildPropertyList(
+      const OperatorDef& operator_def,
+      std::vector<uint64_t>* property_list,
+      std::vector<int64_t>* int_args,
+      std::vector<float>* float_args) {
+    for (const auto& arg: operator_def.arg()) {
+      if (!StartsWith(arg.name(), "custom_")) {
+        continue;
+      }
+      // Pick the name as ABC if the arg name is custom_ABC
+      property_list->push_back((uint64_t)(arg.name().c_str() + 7));
```

Comment:
`property_list` isn't intended for passing arbitrary properties by string key. If you need to pass any non-standard property, define a constant `ONNXIFI_BACKEND_PROPERTY_SOMETHING` for the key.